openedx_external_enrollments/external_enrollments/pathstream_external_enrollment.py

```diff
@@ -0,0 +1,188 @@
+        if self.client is None:
+            raise S3NotInitialized
+
+        try:
+
+            self.client.download_file(
+                self.S3_BUCKET,
+                self.S3_FILE,
+                self.S3_FILE
+            )
+
+        except ClientError as error:
+            LOG.error('Failed to download the file to S3. Reason: %s', str(error))
+            return False
+
+        LOG.info('File [%s] was downloaded from S3 for [%s]', self.S3_FILE, self.__str__())
+        return True
+
+    def _update_file(self, enrollments_data):
+        """Append enrollment data to the downloaded S3 file
+
+        :param enrollments_data: List of the enrollments data
+        """
+        if os.path.exists(self.S3_FILE):
+
+            with open(self.S3_FILE, 'a') as f:
+                f.writelines(enrollments_data)
+        else:
+            LOG.error('File does not exist or _download_file has not been called')
+
+    def _upload_file(self):
+        """Upload the file to an S3 bucket
+
+        :param file_name: File to upload
+        :param bucket: Bucket to upload to
+
+        :return: True if file was uploaded, else False
+        """
+        if self.client is None:
+            raise S3NotInitialized
+
+        try:
+            self.client.upload_file(
+                self.S3_FILE,
+                self.S3_BUCKET,
+                self.S3_FILE
+            )
+
+        except ClientError as error:
+            LOG.error('Failed to upload the file to S3. Reason: %s', str(error))
+            return False
+
+        LOG.info('File [%s] was uploaded to S3 for [%s]', self.S3_FILE, self.__str__())
+        return True
+
+    def _delete_downloaded_file(self):
+        """Delete the downloaded file.
+        """
+        if os.path.exists(self.S3_FILE):
+            os.remove(self.S3_FILE)
+        else:
+            LOG.error('File could not be deleted, File does not exist')
+
+    def execute_upload(self):
+        """
+        Context: This method will download, update and upload an S3 File.
+
+        Warning: Must be called just once per day after midnight
+        """
+
+        enrollments_qs = PathstreamEnrollment.objects.filter(  # pylint: disable=no-member
+            controller_name=str(self),
+            is_uploaded=False,
+        )
+
+        if enrollments_qs.count() > 0:
+
+            self._init_s3()
+
+            is_file_downloaded = self._download_file()
+
+            if is_file_downloaded:
+
+                enrollments_data = [enrollment.meta['enrollment_data_formated'] for enrollment in enrollments_qs]
+
+                self._update_file(enrollments_data)
+
+                is_uploaded = self._upload_file()
+
+                if is_uploaded:
+
+                    enrollments_qs.update(is_uploaded=True)
+
+                self._delete_downloaded_file()
+        else:
+            LOG.info('There are no new enrollments to update the S3 file')
+
+    def _get_enrollment_headers(self):
```

Comment:
@Jacatove, since you are not overriding this method, we can omit it, same for `_get_enrollment_url`.

Reply:
@ivanvgh I was getting `W0223: Method '_get_enrollment_url' is abstract in class 'BaseExternalEnrollment' but is not overridden (abstract-method)`, this is because Pathstream class extends from `BaseExternalEnrollment` but it does not use all its methods. The SDK boto3 provides simplifies that.